packages/react-dom/src/events/DOMEventResponderSystem.js

```diff
@@ -65,161 +66,94 @@ export function setListenToResponderEventTypes(
   listenToResponderEventTypesImpl = _listenToResponderEventTypesImpl;
 }
 
-type EventObjectType = $Shape<PartialEventObject>;
-
-type EventQueue = {
-  events: Array<EventObjectType>,
-  eventPriority: EventPriority,
-};
-
-type PartialEventObject = {
-  target: Element | Document,
-  type: string,
-};
+type EventQueueItem = {|
+  listeners: Array<(val: any) => void>,
+  value: any,
+|};
+type EventQueue = Array<EventQueueItem>;
 
 type ResponderTimeout = {|
   id: TimeoutID,
   timers: Map<number, ResponderTimer>,
 |};
 
 type ResponderTimer = {|
-  isHook: boolean,
-  instance: ReactDOMEventComponentInstance,
+  instance: ReactDOMEventResponderInstance,
   func: () => void,
   id: number,
   timeStamp: number,
 |};
 
 const activeTimeouts: Map<number, ResponderTimeout> = new Map();
-const rootEventTypesToEventComponentInstances: Map<
+const rootEventTypesToEventResponderInstances: Map<
   DOMTopLevelEventType | string,
-  Set<ReactDOMEventComponentInstance>,
+  Set<ReactDOMEventResponderInstance>,
 > = new Map();
-const ownershipChangeListeners: Set<ReactDOMEventComponentInstance> = new Set();
-const PossiblyWeakMap = typeof WeakMap === 'function' ? WeakMap : Map;
-const eventListeners:
-  | WeakMap
-  | Map<
-      $Shape<PartialEventObject>,
-      ($Shape<PartialEventObject>) => void,
-    > = new PossiblyWeakMap();
+const ownershipChangeListeners: Set<ReactDOMEventResponderInstance> = new Set();
 
 let globalOwner = null;
-let continueLocalPropagation = false;
 
 let currentTimeStamp = 0;
 let currentTimers = new Map();
-let currentInstance: null | ReactDOMEventComponentInstance = null;
+let currentInstance: null | ReactDOMEventResponderInstance = null;
 let currentEventQueue: null | EventQueue = null;
+let currentEventQueuePriority: EventPriority = ContinuousEvent;
 let currentTimerIDCounter = 0;
 let currentDocument: null | Document = null;
-let currentlyInHook = false;
 
 const eventResponderContext: ReactDOMResponderContext = {
   dispatchEvent(
-    possibleEventObject: Object,
-    listener: ($Shape<PartialEventObject>) => void,
+    eventProp: string,
+    eventValue: any,
     eventPriority: EventPriority,
   ): void {
     validateResponderContext();
-    const {target, type, timeStamp} = possibleEventObject;
-
-    if (target == null || type == null || timeStamp == null) {
-      throw new Error(
-        'context.dispatchEvent: "target", "timeStamp", and "type" fields on event object are required.',
+    validateEventValue(eventValue);
+    // $FlowFixMe: Flow gets really confused with this line...
```

Comment:
It's because it's an opaque type. Let's make it not an opaque type, if we're going to compare them?

Reply:
Done